disnake/enums.py

```diff
@@ -72,326 +85,1371 @@
+    """
+    Enforces proper enum value typing and prevents duplicate names.
 
-@total_ordering
-class _EnumValueComparable(_EnumValueBase):
-    def __eq__(self, other: object) -> bool:
-        return isinstance(other, self.__class__) and self.value == other.value
+    This is used as the ephemeral namespace in enum class creation.
 
-    def __lt__(self, other: object) -> bool:
-        return isinstance(other, self.__class__) and self.value < other.value
+    Class attributes such as descriptors and private attributes will be directly stored inside this
+    dict, whereas any actual enum members are stored inside ``self.member_map``.
 
+    ``EnumMeta`` will use the collected items in ``self.member_map`` as the enum members.
+    """
 
-def _create_value_cls(name: str, comparable: bool) -> Type[_EnumValueBase]:
-    parent = _EnumValueComparable if comparable else _EnumValueBase
-    return type(f"{parent.__name__}_{name}", (parent,), {"_cls_name": name})  # type: ignore
+    def __init__(self, base: Type[Any]):
+        # We explicitly take a base in __init__, unlike default Enums, to more easily
+        # enforce proper value typing. This should ensure better performance, as e.g.
+        # `Enum[int].one + 3 == 4` is more performant than `Enum[Any].one.value + 3 == 4`,
+        # as this saves a lookup. We do lose some functionality here, but it's probably
+        # worth the cost; especially considering all current enums have monotyped values.
+        super().__init__()
+        self.base = base
+        self.member_map: Dict[str, Any] = {}
+
+    def __setitem__(self, name: str, value: Any) -> None:
+        if name in {
+            # Illegal names defined in original python enums:
+            "mro",
+            "",
+            # Additional illegal names to prevent overwriting:
+            "name",
+            "value",
+        }:
+            # The main Enum must not be checked or its name and value fields will be flagged...
+            if EnumMeta.__is_enum_instantiated__:
+                raise ValueError(f"Invalid Enum member name: {name}")
+
+        if name.startswith("_") or _is_descriptor(value):
+            super().__setitem__(name, value)
+            return
+
+        if not isinstance(value, self.base):
+            raise TypeError(
+                f"Member {name} must be of type {self.base.__name__}, got {type(value).__name__}"
+            )
+
+        if name in self.member_map:
+            raise TypeError(f"Cannot create multiple members with the same name: {name!r}")
+
+        self.member_map[name] = value
 
 
-def _is_descriptor(obj):
-    return hasattr(obj, "__get__") or hasattr(obj, "__set__") or hasattr(obj, "__delete__")
+class EnumMeta(type):
+    __is_enum_instantiated__: ClassVar[bool] = False
+
+    _member_map_: Mapping[str, Enum]
+    _value2member_map_: Mapping[Any, Enum]
+
+    def __new__(
+        metacls: Type[_T],  # pyright: reportSelfClsParameterName=false
```

Comment:
```suggestion
        metacls: Type[_T],  # pyright: ignore[reportSelfClsParameterName]
```
Fairly sure this would otherwise disable `reportSelfClsParameterName` for the entire file.